packages/expo-notifications/android/src/main/java/expo/modules/notifications/tokens/FirebaseTokenListenerService.java

```diff
@@ -0,0 +1,81 @@
+package expo.modules.notifications.tokens;
+
+import com.google.firebase.messaging.FirebaseMessagingService;
+
+import java.lang.ref.WeakReference;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.WeakHashMap;
+
+import androidx.annotation.NonNull;
+
+/**
+ * Subclass of FirebaseMessagingService responsible for dispatching new tokens.
+ */
+public class FirebaseTokenListenerService extends FirebaseMessagingService {
+  // Unfortunately we cannot save state between instances of a service other way
+  // than by static properties. Fortunately, using weak references we can
+  // be somehow sure instances of PushTokenListeners won't be leaked by this component.
+
+  /**
+   * We store this value to be able to inform new listeners of last known token.
+   */
+  private static String LAST_TOKEN = null;
+
+  /**
+   * A collection of weak references to listeners. Used to iterate over, on new token.
+   */
+  private static Collection<WeakReference<PushTokenManager>> LISTENERS = new HashSet<>();
```

Comment:
Can't we just iterate over LISTENERS_REFERENCES map? It would be more readable, imho.